scale/ingest/scan/scanners/scanner.py

```diff
@@ -0,0 +1,317 @@
+from __future__ import unicode_literals
+
+import logging
+import os
+from abc import ABCMeta, abstractmethod
+
+from django.db import transaction
+from django.utils.timezone import now
+
+from ingest.models import Ingest
+from ingest.scan.scanners.exceptions import ScannerInterruptRequested
+from job.configuration.configuration.job_configuration import JobConfiguration, MODE_RW
+from job.configuration.data.job_data import JobData
+from queue.models import Queue
+from storage.media_type import get_media_type
+from storage.models import Workspace
+from trigger.models import TriggerEvent
+from util.file_size import file_size_to_string
+
+logger = logging.getLogger(__name__)
+
+
+class Scanner(object):
+    """Abstract class for a scanner that processes existing files to ingest. Sub-classes must have a no-argument
+    constructor that passes in the correct scanner type and supported broker types and should override the
+    load_configuration(), run(), stop(), and validate_configuration() methods.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, scanner_type, supported_broker_types):
+        """Constructor
+
+        :param scanner_type: The type of this scanner
+        :type scanner_type: string
+        :param supported_broker_types: The broker types supported by this scanner
+        :type supported_broker_types: [string]
+        """
+
+        self.scan_id = None
+        self._count = 0
+        self._dry_run = False  # Used to only scan and skip ingest process
+        self._file_handler = None  # The file handler configured for this scanner
+        self._recursive = True
+        self._scanned_workspace = None  # The workspace model that is being scanned
+        self._scanner_type = scanner_type
+        self._stop_received = False
+        self._supported_broker_types = supported_broker_types
+        self._workspaces = {}  # The workspaces needed by this scanner, stored by workspace name {string: workspace}
+
+    def set_recursive(self, recursive):
+        """Support configuration of scanner recursive property
+        
+        :param recursive: The flag indicating whether workspace will be recursively scanned
+        :type recursive: bool
+        """
+
+        self._recursive = recursive
+
+    @property
+    def scanner_type(self):
+        """The type of this scanner
+
+        :returns: The scanner type
+        :rtype: string
+        """
+
+        return self._scanner_type
+
+    @property
+    def supported_broker_types(self):
+        """The broker types supported by this scanner
+
+        :returns: The broker types supported by this scanner
+        :rtype: [string]
+        """
+
+        return self._supported_broker_types
+
+    @abstractmethod
+    def load_configuration(self, configuration):
+        """Loads the given configuration
+
+        :param configuration: The configuration as a dictionary
+        :type configuration: dict
+        """
+
+        raise NotImplementedError
```

Comment:
BLUF: Both class and instance of are valid forms.
TLDR -->
See PEP 3109 as a reference: `raise EXCEPTION is used to raise a new exception. This form has two sub-variants: EXCEPTION may be an exception class or an instance of an exception class; valid exception classes are BaseException and its subclasses [5]. If EXCEPTION is a subclass, it will be called with no arguments to obtain an exception instance.`